### PR TITLE
doc: add overview of TriggerRawStreamReassembly

### PR DESCRIPTION
**Pull request by inashivb**
Link to ticket: https://redmine.openinfosecfoundation.org/issues/4351

Previous PR: https://github.com/OISF/suricata/pull/11446

Changes since v1:
- Comments addressed
- PR rebased

### CODE REVIEW

File: doc/userguide/devguide/internals/engines/stream/applayertriggerrawreassembly.rst [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+.. image:: stream-engine/trigger_reassembly_after.png
+
+Note that there could be any amount of data in the stream but we're only concerned about the
+last ACK'd data.
+
+Terms displayed in the diagrams:
+
+* ``Base offset``: Offset of the streaming buffer region w.r.t. the actual stream.
+* ``ACK'd data``: The indicator of position in the streaming buffer region till which the last ACK was seen.
+* ``Raw Stream Progress``: This indicates where in the stream the detection engine has already inspected, so it's a "bookmark" of sorts to keep track of where the last inspection left off and from what offset a next call should consider inspection.
```
Comment (jufajardini) [inline doc/userguide/devguide/internals/engines/stream/applayertriggerrawreassembly.rst:40]: nit: line break (we try to keep our documentation character limit per line around 79/80, as indicated by sphinx guidelines :) )

File: doc/userguide/devguide/internals/engines/stream/applayertriggerrawreassembly.rst [hunk 2]
```diff
@@ -0,0 +1,48 @@
+Triggering Raw Stream Reassembly from Application Layer
+-------------------------------------------------------
+
+Function ``AppLayerParserTriggerRawStreamReassembly``
+
+This is a call for immediate reassembly of raw stream data by the Application Layer to make
+more data available for detection e.g. upon completion of an HTTP request as the
+Application Layer may not have access to all the data that has been seen by the stream engine
+at a point in time.
```
Comment (jufajardini) [inline doc/userguide/devguide/internals/engines/stream/applayertriggerrawreassembly.rst:9]: Suggestion:
```suggestion
This is a call by the Application Layer parser for immediate reassembly of the raw stream to make
more data available for detection, e.g. upon completion of an HTTP request as the
Application Layer may not have access to all the data that the stream engine has seen 
at a point in time.
```

File: doc/userguide/devguide/internals/engines/stream/applayertriggerrawreassembly.rst [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+The stream engine provides the following parameters for the stream to be tracked by the different
+subsystems of Suricata.
```
Comment (jufajardini) [inline doc/userguide/devguide/internals/engines/stream/applayertriggerrawreassembly.rst:12]:
```suggestion
subsystems of Suricata:
```

File: doc/userguide/devguide/internals/engines/stream/applayertriggerrawreassembly.rst [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+* app layer progress
+* raw reassembly progress
+* streaming logger progress
+
+These parameters are used to track the progress of the stream in the respective subsystems.
+This serves as a communication between the different subsystems about how far they have
+already looked into the stream. They are all relative to the base offset of the streaming
+buffer region.
+
+For a given streaming buffer region as follows
```
Comment (jufajardini) [inline doc/userguide/devguide/internals/engines/stream/applayertriggerrawreassembly.rst:23]:
```suggestion
For a given streaming buffer region illustrated below:
```

File: doc/userguide/devguide/internals/engines/stream/applayertriggerrawreassembly.rst [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+.. image:: stream-engine/trigger_reassembly_before.png
+
+If a call to ``AppLayerParserTriggerRawStreamReassembly`` was made, things would change as follows
+allowing the Detection Engine to see a bigger part of the stream until the last ACK as seen by the
+stream engine.
```
Comment (jufajardini) [inline doc/userguide/devguide/internals/engines/stream/applayertriggerrawreassembly.rst:29]:
```suggestion
If a call to ``AppLayerParserTriggerRawStreamReassembly`` was made, things would change as represented
in the next image, allowing the Detection Engine to see a bigger part of the stream until the last ACK as seen by the
Stream Engine.
```

**Review by jufajardini — CHANGES_REQUESTED** [review]
It's giving "I'm learning something more in-depth feels, I love that <3

I've left a few suggestions for cases where I felt we could improve readability.

Also realized that in some places we're capitalizing the engine names (e.g. Detection Engine), whereas in others we're not. I think it's better if we use capitals in all places. :)

File: doc/userguide/devguide/internals/engines/stream/applayertriggerrawreassembly.rst [hunk 6]
```diff
@@ -0,0 +1,48 @@
+Triggering Raw Stream Reassembly from Application Layer
+-------------------------------------------------------
+
+Function ``AppLayerParserTriggerRawStreamReassembly``
```
Comment (inashivb) [inline doc/userguide/devguide/internals/engines/stream/applayertriggerrawreassembly.rst:4]: This looks a bit odd to me. Do you have any suggestions to incorporate the fn name better @jufajardini ?

File: doc/userguide/devguide/internals/engines/stream/applayertriggerrawreassembly.rst [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+This is a call for immediate reassembly of raw stream data by the Application Layer to make
+more data available for detection e.g. upon completion of an HTTP request as the
+Application Layer may not have access to all the data that has been seen by the stream engine
```
Comment (inashivb) [inline doc/userguide/devguide/internals/engines/stream/applayertriggerrawreassembly.rst:8]: There's more to this. Needs rework.

File: doc/userguide/devguide/internals/engines/stream/applayertriggerrawreassembly.rst (hunk 6 above)
Comment (jufajardini) [reply doc/userguide/devguide/internals/engines/stream/applayertriggerrawreassembly.rst:4]: How about:

It is possible to trigger the reassembly of the raw stream from the application layer protocol parsers, by calling the function ``AppLayerParserTriggerRawStreamReassembly``.

(then proceed with the rest)?

**Pull request closed without merging**